sw/airborne/firmwares/rotorcraft/guidance/guidance_hybrid.c

```diff
@@ -0,0 +1,397 @@
+    int32_t cv = INT_MULT_RSHIFT(wind_estimate.x, wind_estimate.x, 8) + INT_MULT_RSHIFT(wind_estimate.y, wind_estimate.y, 8) - (max_airspeed<<8)*max_airspeed;
+
+    float dv = POS_FLOAT_OF_BFP(bv) * POS_FLOAT_OF_BFP(bv) - 4.0* POS_FLOAT_OF_BFP(av) * POS_FLOAT_OF_BFP(cv);
+    float d_sqrt_f = sqrtf(dv);
+    int32_t d_sqrt = POS_BFP_OF_REAL(d_sqrt_f);
+
+    int32_t result = ((-bv + d_sqrt)<<8)/(2*av);
+
+    guidance_hybrid_airspeed_sp.x = wind_estimate.x + INT_MULT_RSHIFT(guidance_hybrid_groundspeed_sp.x, result, 8);
+    guidance_hybrid_airspeed_sp.y = wind_estimate.y + INT_MULT_RSHIFT(guidance_hybrid_groundspeed_sp.y, result, 8);
+  }
+  else {
+    // Add the wind to get the airspeed setpoint
+    guidance_hybrid_airspeed_sp = guidance_hybrid_groundspeed_sp;
+    VECT2_ADD(guidance_hybrid_airspeed_sp, wind_estimate);
+  }
+
+//   limit the airspeed setpoint to 15 m/s, because else saturation+windup will occur
+  norm_airspeed_sp = int32_vect2_norm(&guidance_hybrid_airspeed_sp);
+  if(norm_airspeed_sp > (max_airspeed<<8)) {
+    guidance_hybrid_airspeed_sp.x = guidance_hybrid_airspeed_sp.x*(max_airspeed<<8)/norm_airspeed_sp;
+    guidance_hybrid_airspeed_sp.y = guidance_hybrid_airspeed_sp.y*(max_airspeed<<8)/norm_airspeed_sp;
+  }
+}
+
+void guidance_hybrid_determine_wind_estimate(void) {
+
+  /* compute speed error    */
+  struct Int32Vect2 wind_estimate_measured;
+  struct Int32Vect2 measured_ground_speed;
+  INT32_VECT2_RSHIFT(measured_ground_speed, *stateGetSpeedNed_i(), 11);
+  VECT2_DIFF(wind_estimate_measured, guidance_hybrid_ref_airspeed, measured_ground_speed );
+
+  //Low pass wind_estimate, because we know the wind usually only changes slowly
+  //But not too slow, because the wind_estimate is also an adaptive element for the airspeed model inaccuracies
+  wind_estimate_high_res.x += (( (wind_estimate_measured.x - wind_estimate.x) > 0)*2-1) * wind_gain;
+  wind_estimate_high_res.y += (( (wind_estimate_measured.y - wind_estimate.y) > 0)*2-1) * wind_gain;
+
+  wind_estimate.x = ((wind_estimate_high_res.x) >> 8);
+  wind_estimate.y = ((wind_estimate_high_res.y) >> 8);
+}
+
+void guidance_hybrid_set_cmd_i(struct Int32Eulers *sp_cmd) {
+  /// @todo calc sp_quat in fixed-point
+
+  /* orientation vector describing simultaneous rotation of roll/pitch */
+  struct FloatVect3 ov;
+  ov.x = ANGLE_FLOAT_OF_BFP(sp_cmd->phi);
+  ov.y = ANGLE_FLOAT_OF_BFP(sp_cmd->theta);
+  ov.z = 0.0;
+  /* quaternion from that orientation vector */
+  struct FloatQuat q_rp;
+  FLOAT_QUAT_OF_ORIENTATION_VECT(q_rp, ov);
+  struct Int32Quat q_rp_i;
+  QUAT_BFP_OF_REAL(q_rp_i, q_rp);
+
+  //   get the vertical vector to rotate the roll pitch setpoint around
+  struct Int32Vect3 zaxis = {0, 0, 1};
+
+  /* get current heading setpoint */
+  struct Int32Quat q_yaw_sp;
+  INT32_QUAT_OF_AXIS_ANGLE(q_yaw_sp, zaxis, sp_cmd->psi);
```

Comment:
please use the functions instead of the macros...
We only kept the macros for backwards compatibility and would like to remove them at some point.